### ISSUE TITLE
Input region only gets updated after moving mouse

### ISSUE THREAD

**Issue by fwSmit** [swaywm/sway#5911]
I'm working on getting [wayland support for dunst](https://github.com/dunst-project/dunst/pull/781), using lots of code from mako. There's one bug that I can't seem to fix. We use layer_surfaces to display notifications. One surface is used for multiple notifications and is resized when a notification is dismissed. The problem is that the input region doesn't get immediately updated on resizing the surface. If you click another time, seemingly outside the surface, the click still goes to the dunst window. Only if you move your mouse slightly after clicking, the next click goes to the window below.
It seems like this is not a dunst bug, because a similar thing happens on firefox. When you have autoscroll enabled, middleclick to use autoscroll. If you move your mouse and click on the round thing that appeared, nothing happens. The click is absorbed. Only if you don't move your mouse, the click goes to the firefox window below (note on x11, clicking the round thing makes it go away as well. On wayland not (yet?)).  
This is tested on sway and hikari.

**Comment by RPigott** [comment]
I don't have more info about the reported issue, but I believe the firefox issue is a firefox bug. It's reproducible for me on both sway and weston, and I think mutter as well based on reports, though I haven't tested it myself.

First appeared in FF71: https://bugzilla.mozilla.org/show_bug.cgi?id=1565313#c8
Regressed even more in FF77 w/ this patch: https://bugzilla.mozilla.org/show_bug.cgi?id=1623974 , now the widget crashes both sway (#5321) and FF (https://bugzilla.mozilla.org/show_bug.cgi?id=1661192#c7)
Fixed in FF83 to return the FF71 behavior: https://bugzilla.mozilla.org/show_bug.cgi?id=1661192#c15

It could also possibly be a gtk3 bug.

**Comment by fwSmit** [comment]
> I don't have more info about the reported issue, but I believe the firefox issue is a firefox bug. It's reproducible for me on both sway and weston, and I think mutter as well based on reports, though I haven't tested it myself.
> 
> First appeared in FF71: https://bugzilla.mozilla.org/show_bug.cgi?id=1565313#c8
> Regressed even more in FF77 w/ this patch: https://bugzilla.mozilla.org/show_bug.cgi?id=1623974 , now the widget crashes both sway (#5321) and FF (https://bugzilla.mozilla.org/show_bug.cgi?id=1661192#c7)
> Fixed in FF83 to return the FF71 behavior: https://bugzilla.mozilla.org/show_bug.cgi?id=1661192#c15


Yeah it's annoying that firefox takes so long to fix wayland bugs. There are a few that are a couple years old and still haven't been fixed.
What I meant however was related, but not entirely the same. When middle clicking, the autoscroll widget appears. Without moving the mouse, click again and the widget disappears (which may be caused by sway). When the mouse is moved before clicking, but still inside the widget, the widget doesn't disappear (which is the bug you linked).

I can't reproduce the sway bug in firefox on my other machine. I can reproduce the dunst bug there.

> It could also possibly be a gtk3 bug.

Dunst doesn't use gtk3. It could still be due to a bad dunst implementation, but I haven't found any issues on that side. I even tried manually changing the input region on resizing the window, but that didn't have any effect.

**Comment by RPigott** [comment]
> Without moving the mouse, click again and the widget disappears

I hadn't noticed that difference, but you're right. I wonder if that's new in FF83 compared to FF71. I guess I didn't carefully verify that their behavior was the same – the autoscroll widget and webrender compositor have been pretty buggy on Wayland for a while.

I'll definitely look into this more in sway. I took me a while to explain the crashes in firefox and I wasn't looking forward to doing that again.

> > It could also possibly be a gtk3 bug.
>
> Dunst doesn't use gtk3. 

Yeah, I meant the (possible) firefox bug. If they are the same issue in sway having two clients to test is great news.

**Comment by fwSmit** [comment]
I'm not sure how much you know about wayland, but a feature of wayland is that clients only see user input when in focus. If the x11 implementation the autoscroll widget wasn't transparent to input, it still could work.



> > Without moving the mouse, click again and the widget disappears
> 
> I hadn't noticed that difference, but you're right. I wonder if that's new in FF83 compared to FF71. I guess I didn't carefully verify that their behavior was the same – the autoscroll widget and webrender compositor have been pretty buggy on Wayland for a while.

So you can reproduce the sway bug in Firefox? As I said, I can only reproduce it on half my machines.

**Comment by RPigott** [comment]
> I'm not sure how much you know about wayland, but a feature of wayland is that clients only see user input when in focus. If the x11 implementation the autoscroll widget wasn't transparent to input, it still could work.

Yes. And as I'm sure you know, in Wayland clients may manage multiple surfaces and pointer input is tracked per-surface with enter/leave messages. Firefox is a somewhat unusual client in that it creates a popup surface and then creates a subsurface on that popup to actually draw the autoscroll widget. Firefox's input code expects to consume input events from gdk, which gdk generates from wayland events related to the popup surface it manages. From my previous work with firefox, I know that FF at least tries to do the right thing by setting an empty input region on the subsurface and configuring gdk to pass on the events it receives.

> So you can reproduce the sway bug in Firefox? As I said, I can only reproduce it on half my machines.

Previously I had never seen the widget dismissed by clicking within it, motion or not. When you said you were able to dismiss the widget in firefox without pointer motion, but it was device dependent, based on my previous investigation of firefox I realized there must be a sway bug.

With those hints, I am now able to reproduce the issue as you describe. If I middle-click with my mouse to summon the widget, I cannot dismiss it. If I three-finger tap with my _touchpad_ to summon the widget, I can dismiss the widget with a click – provided my pointer has not moved. See if that is true on your machines as well.

Firefox cannot distinguish these inputs, so one of these cases is bugged in sway. Sadly, I think its the second case. So firefox is also bugged, and fixing this issue in sway will actually make the firefox behavior worse. Hopefully we can then correct firefox's behavior to allow dismissing the widget by clicking on it.

**Comment by fwSmit** [comment]
> Yes. And as I'm sure you know, in Wayland clients may manage multiple surfaces and pointer input is tracked per-surface with enter/leave messages. Firefox is a somewhat unusual client in that it creates a popup surface and then creates a subsurface on that popup to actually draw the autoscroll widget. Firefox's input code expects to consume input events from gdk, which gdk generates from wayland events related to the popup surface it manages. From my previous work with firefox, I know that FF at least tries to do the right thing by setting an empty input region on the subsurface and configuring gdk to pass on the events it receives.

Hmm interesting, I didn't know that.

> With those hints, I am now able to reproduce the issue as you describe. If I middle-click with my mouse to summon the widget, I cannot dismiss it. If I three-finger tap with my _touchpad_ to summon the widget, I can dismiss the widget with a click – provided my pointer has not moved. See if that is true on your machines as well.

You're right! It only works when three finger clicking on a touchpad. When middle clicking by pressing in the mid bottom of the touchpad it doesn't work either. That's very weird.

> Firefox cannot distinguish these inputs, so one of these cases is bugged in sway. Sadly, I think its the second case. So firefox is also bugged, and fixing this issue in sway will actually make the firefox behavior worse. Hopefully we can then correct firefox's behavior to allow dismissing the widget by clicking on it.

I think it would be easiest to fix the firefox behaviour by making the widget transparent to input. Why don't we use that solution?

**Comment by RPigott** [comment]
> You're right! It only works when three finger clicking on a touchpad. When middle clicking by pressing in the mid bottom of the touchpad it doesn't work either. That's very weird.

Great! Yeah, it's very unusual behavior. I would not normally have guessed to try that.

> I think it would be easiest to fix the firefox behaviour by making the widget transparent to input. Why don't we use that solution?

So far I'm just guessing at what is wrong. The mouse/touchpad behavior we observed is likely a sway bug, so it should be addressed. When we are certain sway is doing the right thing, if Firefox still has bad behavior Firefox can be fixed too.

**Comment by RPigott** [comment]
I've looked into this a bit more and the behavior is timing based.

For me, if the middle button is depressed for less than ~25ms the widget is dismissable. If it is depressed for longer the widget is undismissable.

I've been using a wlr-virtual-pointer client to carefully time the button events, e.g
```c
zwlr_virtual_pointer_v1_button(vptr, timestamp(), button, WL_POINTER_BUTTON_STATE_PRESSED);
zwlr_virtual_pointer_v1_frame(vptr);
wl_display_flush(display);
usleep(25 * 1000);
zwlr_virtual_pointer_v1_button(vptr, timestamp(), button, WL_POINTER_BUTTON_STATE_RELEASED);
zwlr_virtual_pointer_v1_frame(vptr);
```

The Wayland timestamp of the press and release events from the libinput touchpad driver are ~50ms apart according to wev, but it still allows me to dismiss the widget. The actual time delta between when these are seen by wev is smaller though. When I use the virtual pointer the actual time delta is much closer to the sleep amount.

This would normally lead me to believe this is a race in firefox, but I can't reproduce in weston with my touchpad and firefox appears to be doing the right thing at a glance.

**Comment by fwSmit** [comment]
> I've looked into this a bit more and the behavior is timing based.
> 
> For me, if the middle button is depressed for less than ~25ms the widget is dismissable. If it is depressed for longer the widget is undismissable.
> 
> I've been using a wlr-virtual-pointer client to carefully time the button events, e.g
> 
> ```c
> zwlr_virtual_pointer_v1_button(vptr, timestamp(), button, WL_POINTER_BUTTON_STATE_PRESSED);
> zwlr_virtual_pointer_v1_frame(vptr);
> wl_display_flush(display);
> usleep(25 * 1000);
> zwlr_virtual_pointer_v1_button(vptr, timestamp(), button, WL_POINTER_BUTTON_STATE_RELEASED);
> zwlr_virtual_pointer_v1_frame(vptr);
> ```
> 
> The Wayland timestamp of the press and release events from the libinput touchpad driver are ~50ms apart according to wev, but it still allows me to dismiss the widget. The actual time delta between when these are seen by wev is smaller though. When I use the virtual pointer the actual time delta is much closer to the sleep amount.
> 
> This would normally lead me to believe this is a race in firefox, but I can't reproduce in weston with my touchpad and firefox appears to be doing the right thing at a glance.

That's interesting. If it's a race condition, then it's a very reliable one.

I've made debug traces from the firefox case. It seems like the pointer only enters a surface once it moves. Not if the surface resizes under the cursor (and probably not it the surface move under the cursor either).

https://gist.github.com/fwSmit/5935f3058e2987cf4942d80fe85cec9e#file-firefox-middle-click-with-move-L5297

Middle click, move cursor a small amount and click again.

https://gist.github.com/fwSmit/4057285a5b7f50f4b7fa480243c1f32c#file-firefox-middle-click-without-move-L6351

Middle click and click again.

I've linked to where the middle click starts.

Traces are made with:
```fish
WAYLAND_DEBUG=1 MOZ_ENABLE_WAYLAND=1 firefox |& tee debug.out
```
